numpyro/util.py

```diff
@@ -270,14 +275,369 @@ def update_fn(t, accept_prob, z_flat, state):
+    # If we are going to the right, start from the right leaf of the current tree.
+    z, r, z_grad = _get_leaf(current_tree, going_right)
+
+    # Then build a new tree.
+    if iterative_build:
+        new_tree = _iterative_build_subtree(current_tree.depth, vv_update, kinetic_fn,
+                                            z, r, z_grad, inverse_mass_matrix, step_size,
+                                            going_right, key,
+                                            energy_current, slice_exp_term, max_sliced_energy,
+                                            use_multinomial_sampling, max_tree_depth)
+    else:
+        new_tree = _build_subtree(current_tree.depth, vv_update, kinetic_fn, z, r, z_grad,
+                                  inverse_mass_matrix, step_size, going_right, key,
+                                  energy_current, slice_exp_term, max_sliced_energy,
+                                  use_multinomial_sampling)
+    return _combine_tree(current_tree, new_tree, inverse_mass_matrix, going_right, transition_key,
+                         use_multinomial_sampling, get_transition_prob, iterative_build=False)
+
+
+def _leaf_idx_to_ckpt_idx(n):
+    # computes the number of non-zero bits except the last bit
+    # e.g. 6 -> 2, 7 -> 2, 13 -> 2
+    _, idx_max = lax.while_loop(lambda nc: nc[0] > 0,
+                                lambda nc: (nc[0] >> 1, nc[1] + (nc[0] & 1)),
+                                (n >> 1, 0))
+    # computes the number of last non-zero bits
+    # e.g. 6 -> 0, 7 -> 3, 13 -> 1
+    _, num_subtrees = lax.while_loop(lambda nc: (nc[0] & 1) != 0,
+                                     lambda nc: (nc[0] >> 1, nc[1] + 1),
+                                     (n, 0))
+    idx_min = idx_max - num_subtrees + 1
+    return idx_min, idx_max
+
+
+def _is_iterative_turning(leaf_idx, inverse_mass_matrix, r, r_sum, r_ckpts, r_sum_ckpts):
+    r, _ = ravel_pytree(r)
+    r_sum, _ = ravel_pytree(r_sum)
+
+    ckpt_idx_min, ckpt_idx_max = _leaf_idx_to_ckpt_idx(leaf_idx)
+    # we update checkpoints when leaf_idx is even
+    r_ckpts, r_sum_ckpts = lax.cond(leaf_idx % 2 == 1,
+                                    (r_ckpts, r_sum_ckpts),
+                                    lambda x: x,
+                                    (r_ckpts, r_sum_ckpts),
+                                    lambda x: (index_update(x[0], ckpt_idx_max, r),
+                                               index_update(x[1], ckpt_idx_max, r_sum)))
+
+    def _body_fn(state):
+        i, _ = state
+        subtree_r_sum = r_sum - r_sum_ckpts[i] + r_ckpts[i]
+        # XXX we might not need to unravel here
+        return i - 1, _is_turning(inverse_mass_matrix, r_ckpts[i], r, subtree_r_sum)
+
+    _, turning = lax.while_loop(lambda it: (it[0] >= ckpt_idx_min) & ~it[1],
+                                _body_fn,
+                                (ckpt_idx_max, False))
+    return turning, r_ckpts, r_sum_ckpts
```

Comment:
Nice! It will be great to have a simple test for this.

Reply:
Yes, I'll add it. How about we use jax master version for a while before starting pinning to a specific version (e.g. next jax release)?

Reply:
I think at least for CI, it will be good to pin some version from master and host it on our end, but for local development, we can just use jax master. I think we shouldn't block ourselves on the next jax release.